File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js

const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
let input = fs.readFileSync(filePath).toString().trim().split("\n");

let [n, m, s] = input[0].split(" ").map((e) => parseInt(e));
let graph = new Array(n + 1);
for (let i = 0; i < graph.length; i++) {
  graph[i] = [];
}
for (let i = 0; i < m; i++) {
  let [from, to] = input[i + 1].split(" ").map((e) => parseInt(e));
  graph[from].push(to);
  graph[to].push(from);
}

graph.forEach((e) => {
  e.sort((a, b) => a - b);
});

const bfs = (g, startNode) => {
  const visited = [];
  let needVisit = [];

  needVisit.push(startNode);

  while (needVisit.length !== 0) {
    const curNode = needVisit.shift();
    if (!visited.includes(curNode)) {
      visited.push(curNode);
      needVisit = [...needVisit, ...g[curNode]]; //needVisit부터 다 하고 (너비 탐색)
    }
  }
  return visited.join(" ");
};

const dfs = (g, startNode) => {
  const visited = [];
  let needVisit = [];

  needVisit.push(startNode);

  while (needVisit.length !== 0) {
    const curNode = needVisit.shift();
    if (!visited.includes(curNode)) {
      visited.push(curNode);
      needVisit = [...g[curNode], ...needVisit]; //needVisit이 후순위로 밀려난다 (깊이 탐색)
    }
  }
  return visited.join(" ");
};

console.log(dfs(graph, s));
console.log(bfs(graph, s));

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

input = sys.stdin.readline

n, m, v = map(int, input().split())

g = [[0 for i in range(n+1)] for j in range(n+1)]

for _ in range(m):
    a, b = map(int, input().split())
    g[a][b] = 1
    g[b][a] = 1

def bfs(v):
    visited = [0 for _ in range(n + 1)]
    result = []

    q = deque()

    visited[v] = 1
    q.append(v)

    while len(q):
        cur = q.popleft()
        result.append(cur)

        for i in range(1,n+1):
            if visited[i] == 0 and g[cur][i] == 1:
                visited[i] = 1
                q.append(i)

    return print(*result)

def dfs(v):
    visited = [0] * (n+1)
    result = []

    def dfs_recursive(node):
        visited[node] = 1
        result.append(node)

        for i in range(1, n+1):
            if visited[i] == 0 and g[node][i] == 1:
                dfs_recursive(i)

    dfs_recursive(v)
    return print(*result)

dfs(v)
bfs(v)


